src/libsync/propagatecommonzsync.cpp

```diff
@@ -0,0 +1,320 @@
+       return stream_error("fwrite", fout);
+    }
+}
+
+/* write_block_sums(buffer[], num_bytes, output_stream)
+ * Given one block of data, calculate the checksums for this block and write
+ * them (as raw bytes) to the given output stream */
+void ZsyncGenerateJob::write_block_sums(unsigned char *buf, size_t got, FILE * f) {
+    struct rsum r;
+    unsigned char checksum[CHECKSUM_SIZE];
+
+    /* Pad for our checksum, if this is a short last block  */
+    if (got < _blocksize)
+        memset(buf + got, 0, _blocksize - got);
+
+    /* Do rsum and checksum, and convert to network endian */
+    r = rcksum_calc_rsum_block(buf, _blocksize);
+    rcksum_calc_checksum(&checksum[0], buf, _blocksize);
+    r.a = htons(r.a);
+    r.b = htons(r.b);
+
+    /* Write them raw to the stream */
+    if (fwrite(&r, sizeof r, 1, f) != 1)
+       return stream_error("fwrite", f);
+    if (fwrite(checksum, sizeof checksum, 1, f) != 1)
+       return stream_error("fwrite", f);
+}
+
+/* read_stream_write_blocksums(data_stream, zsync_stream)
+ * Reads the data stream and writes to the zsync stream the blocksums for the
+ * given data. No compression handling.
+ */
+void ZsyncGenerateJob::read_stream_write_blocksums(FILE * fin, FILE * fout) {
+    unsigned char *buf = (unsigned char *) malloc(_blocksize);
+
+    if (!buf) {
+        fprintf(stderr, "out of memory\n");
+        exit(1);
+    }
+
+    while (!feof(fin)) {
+        int got = fread(buf, 1, _blocksize, fin);
+
+        if (got > 0) {
+            /* The SHA-1 sum, unlike our internal block-based sums, is on the whole file and nothing else - no padding */
+            SHA1Update(&_shactx, buf, got);
+
+            write_block_sums(buf, got, fout);
+            _len += got;
+        }
+        else {
+            if (ferror(fin))
+               return stream_error("fread", fin);
+        }
+    }
+    free(buf);
+}
+
+void ZsyncGenerateJob::generate()
+{
+    // Create a temporary file to use with zsync_begin()
+    QTemporaryFile zsynctf;
+    _zsyncFile.open();
+    zsynctf.open();
+    int fileHandle = _zsyncFile.handle();
+    int tfHandle = zsynctf.handle();
+
+    FILE* out = fdopen(dup(fileHandle), "w");
+    setvbuf(out, NULL, _IONBF, 0);
+    FILE* tf = fdopen(dup(tfHandle), "w+");
+    setvbuf(tf, NULL, _IONBF, 0);
+    std::string fileString = _file.toLocal8Bit().constData();
+    FILE *in = fopen(fileString.c_str(), "r");
+    setvbuf(in, NULL, _IONBF, 0);
+
+    int rsum_len, checksum_len, seq_matches;
+    qCDebug(lcZsyncGenerate) << "Starting generation of:" << _file;
+
+    if (!in) {
+        qCWarning(lcZsyncGenerate) << "Failed to open input file:" << _file;
+        fclose(tf);
+        fclose(out);
+
+        deleteLater();
+
+        emit failedSignal();
+        return;
+    }
+
+    /* Read the input file and construct the checksum of the whole file, and
+     * the per-block checksums */
+    SHA1Init(&_shactx);
+    read_stream_write_blocksums(in, tf);
```

Comment:
I find it surprising that there would be no zsync function for this.

Reply:
There isn't, `read_stream_write_blocksums` is part of `make.c` which is a zsync application, not part of the exported api/header.

Reply:
Hm, I'd *much* prefer if this was part of the library and we wouldn't import large complicated blocks of its code into our codebase. I'll need to read more of this, but maybe it wouldn't be extraordinarily hard to make these kinds of zsync functions available. (is the author receptive to PRs?)

Reply:
I have no problem modifying zsync to export these functions instead, will need to work out how to handle stream_error calls in that case though ...

Reply:
CC @cph6